src/pipert2/core/base/wires_validator.py

```diff
@@ -0,0 +1,90 @@
+    """Validate wires by logical rules: contain source and destination,
+                                        link source to middle/end and source/middle to end,
+                                        each middle routine is consumed and produced by another routines.
+
+        Throws a pipe_validation_failed if validation failed.
+
+    Args:
+        wires: Wires to validate.
+    """
+
+    validate_existing_source_and_destination_routines(wires)
+    validate_routines_place_properly(wires)
+    validate_consume_and_produce_on_middle_routines(wires)
+
+
+def validate_routines_place_properly(wires: List[Wire]):
+    """Validate the source and destination routines - Raise a PipeValidationError if source placed as middle or
+    destination routine placed as source routine.
+
+    Args:
+        wires: Wires to validate.
+    """
+
+    for wire in wires:
+        if isinstance(wire.source, DestinationRoutine):
+            raise PipeValidationError("Can't link any destination routine to any routine")
+        else:
+            for destination_routine in wire.destinations:
+                if isinstance(destination_routine, SourceRoutine):
+                    raise PipeValidationError("Can't link any routine to source routine")
+
+
+def validate_existing_source_and_destination_routines(wires: List[Wire]):
+    """Validate existing source routine and destination routine - Raise a PipeValidationError if doesn't exist.
+
+    Args:
+        wires: Wires to validate.
+    """
+
+    source_routine_exists = False
+    destination_routine_exists = False
+
+    for wire in wires:
+        if isinstance(wire.source, SourceRoutine):
+            source_routine_exists = True
+
+        for destination_routine in wire.destinations:
+            if isinstance(destination_routine, DestinationRoutine):
+                destination_routine_exists = True
+
+    if not source_routine_exists:
+        raise PipeValidationError("Source routine doesn't exist")
+    if not destination_routine_exists:
+        raise PipeValidationError("Destination routine doesn't exist")
+
+
+def validate_consume_and_produce_on_middle_routines(wires: List[Wire]):
+    """Validate each middle routine consume data from another routine and consumed by another routine.
+    Raise a PipeValidationError if doesn't consume or consumed by.
+
+    Args:
+        wires: Wires to validate.
+    """
+
+    middle_routines = {}
+
+    for wire in wires:
+        if isinstance(wire.source, MiddleRoutine):
+            middle_routines.setdefault(wire.source.name, {"source": True, "destination": False})
+            middle_routines[wire.source.name].update({"source": True})
+
+        for destination_routine in wire.destinations:
+            if isinstance(destination_routine, MiddleRoutine):
+                middle_routines.setdefault(destination_routine.name, {"destination": True, "source": False})
+                middle_routines[destination_routine.name].update({"destination": True})
```

Comment:
The lines
`            middle_routines.setdefault(wire.source.name, {"source": True, "destination": False})
            middle_routines[wire.source.name].update({"source": True})`

And 
`                middle_routines.setdefault(destination_routine.name, {"destination": True, "source": False})
                middle_routines[destination_routine.name].update({"destination": True})`

Should switch their places, because for example:
when you are holding the wire.source routine and it is a MiddleRoutine, the routine is the source in this link which means she has a destination and not a source.